feat: Support for .dde.yml Configuration File and Flexible Container Shell

sbaerlocher opened the pull request:
### Added
- **_loadProjectDotdde Function for .dde.yml File Processing**:
  - Implements a new function to check and read configurations from a `.dde.yml` file within the current directory.
  - Supports reading of various settings such as version, container shell, and database configurations.

- **Integration of .dde.yml Configurations in Shell Scripts**:
  - Includes the `_loadProjectDotdde` function in `exec.sh`, `up.sh`, and `update.sh` to load and apply settings from `.dde.yml`.
  - Enhances script functionality to adapt to dynamic configurations provided in the `.dde.yml` file.

- **Flexible Container Shell Configuration via .dde.yml**:
  - Introduces `DDE_CONTAINER_SHELL` environment variable to dynamically set the container's shell based on `.dde.yml`.
  - Adapts container environment to use the specified shell in the `.dde.yml` file, defaulting to `sh` if not specified.

- **Modifications in `configure-image.sh` for Shell Configuration**:
  - Updates the script to handle new shell configurations from `.dde.yml`.
  - Installs and sets up the specified shell for the user, enhancing compatibility across different containers.

### Modified
- **Enhanced Error Handling and Backward Compatibility for .dde.yml File**:
  - Adds robust error checking and messages for missing or incorrect configurations in the `.dde.yml` file.
  - Ensures backward compatibility by skipping `.dde.yml` processing if the file does not exist, maintaining current functionality for projects without a `.dde.yml` file.

- **Version Support in .dde.yml**:
  - Implements support for different versions of `.dde.yml`, allowing for future expansion and modifications.

### Configuration File Example
```yaml
---
version: "1"

container:
  shell: zsh

databases:
  type: postgres
  names: postgres
---

Commit:
feat: add project .dde file

s4mpl3d commented on commands/_internals/loadProjectDotdde.sh:
```diff
@@ -0,0 +1,32 @@
+                export DDE_CONTAINER_SHELL
+
+                # Read and process the 'databases.type' value from the .dde.yml file
+                DDE_DATABASE_TYPE=$(_yq_stdin e '.databases.type' < "$DDE_FILE")
```
@sbaerlocher this is currently unused. Best add these variables when functionality is also added.

s4mpl3d commented on commands/_internals/loadProjectDotdde.sh:
```diff
@@ -0,0 +1,32 @@
+                DDE_DATABASE_TYPE=$(_yq_stdin e '.databases.type' < "$DDE_FILE")
+
+                # Process the list of database names under 'databases.names'
+                DDE_DATABASE_NAMES=$(_yq_stdin e '.databases.names[]' < "$DDE_FILE")
```
@sbaerlocher same

s4mpl3d commented on commands/_internals/loadProjectDotdde.sh:
```diff
@@ -0,0 +1,32 @@
+        DDE_VERSION=$(_yq_stdin e '.version' < "$DDE_FILE")
+
+        case $DDE_VERSION in
+            1)
```
@sbaerlocher Can we version this more fluently? It should be compatible with semver and allow us to only override whats changed, without redefining everything

s4mpl3d commented on dde.sh:
```diff
@@ -14,9 +14,10 @@ DOCKER_BUILDKIT=1
 DDE_UID=$(id -u)
 DDE_GID=$(id -g)
 DDE_BROWSER=
+DDE_CONTAINER_SHELL=sb
```
@sbaerlocher typo, should probably be sh

s4mpl3d commented on helper/configure-image.sh:
```diff
@@ -1,20 +1,23 @@
-###############################################################################
-
-# Check plaform
+DEFAULT_SHELL="sb"
```
@sbaerlocher typo. Can't we define the default globaly?

s4mpl3d commented on helper/configure-image.sh:
```diff
@@ -1,20 +1,23 @@
-# Check plaform
+DEFAULT_SHELL="sb"
+# Set default shell to /bin/sb, can be overridden by setting DDE_SHELL
+SHELL_TO_INSTALL="/bin/${DDE_CONTAINER_SHELL:-$DEFAULT_SHELL}"
```
@sbaerlocher how does DDE_CONTAINER_SHELL get into the container? If a configuration/image change is needed it should be documented

s4mpl3d commented on helper/configure-image.sh:
```diff
@@ -24,26 +27,34 @@ else
+    apt-get)
+        apt-get update
+        apt-get install -qq curl
+        [ "$SHELL_TO_INSTALL" = "/bin/zsh" ] && apt-get install -y zsh
```
@sbaerlocher this only works with zsh, bash should also be installed when chosen in the config file. We could just reuse $DDE_CONTAINER_SHELL

s4mpl3d replied:
@sbaerlocher why not directly use $DEFAULT_SHELL if set? Performance wise it would be better if we call apk add only once

s4mpl3d commented on helper/configure-image.sh:
```diff
@@ -24,26 +27,34 @@ else
+useradd -d /home/dde -u $DDE_UID -g $DDE_GID -c "dde" -s "$SHELL_TO_INSTALL" -N -o -m dde -r || true
+
+# Install oh-my-zsh for the dde user if zsh is the selected shell
+if [ "$SHELL_TO_INSTALL" = "/bin/zsh" ] && which zsh > /dev/null 2>&1; then
```
@sbaerlocher not everyone choosing zsh as default shell will want to use oh my zsh. I would remove this and add it into the project build if needed

s4mpl3d commented on README.md:
````diff
@@ -190,6 +190,26 @@ if you use `volume` you must expose the volume in the `docker-compose.override.y
+To configure a custom shell in whatwedo/dde using the .dde.yml configuration file, specify your preferred shell with the container.shell key. This setting instructs whatwedo/dde to use the specified shell within the container. For instance, to utilize zsh as the container shell, your configuration would appear as follows:
+
+```yml
+version: "1"
````
@sbaerlocher this value will result in "Unsupported version ($DDE_VERSION) in .dde.yml. Using default settings."

sbaerlocher committed:
feat: add SemVer Version

docs: add doc custom shell

docs: add docker compose

feat: bash support

fix: remove ohmyzsh

fix: compare

s4mpl3d commented on README.md:
```diff
@@ -190,6 +190,26 @@ if you use `volume` you must expose the volume in the `docker-compose.override.y
 
 The environment variables `DDE_DNS_FORWARD_1` and `DDE_DNS_FORWARD_2` allow for setting custom DNS servers in the Docker Development Environment (DDE). This is useful when local Internet DNS servers are to be used.
 
+### Configuring a Custom Shell
```
@sbaerlocher as discussed,  this is currently in the middle of the dns documentation. Also the argument in the Dockerfile should be added to the documentation

sbaerlocher committed:
fix: docu

s4mpl3d commented on helper/configure-image.sh:
```diff
@@ -1,20 +1,24 @@
+# Check if necessary environment variables are set
+[ -z "$DDE_UID" ] && echo "DDE_UID is not set" && exit 1
+[ -z "$DDE_GID" ] && echo "DDE_GID is not set" && exit 1
+[ -z "$DDE_CONTAINER_SHELL" ] && echo "DDE_CONTAINER_SHELL is not set" && exit 1
```
@sbaerlocher why exit here if you define another default a couple lines down? Also we got multiple default values in exec.sh:30+31, dde.sh:17

sbaerlocher replied:
The line is removed, and within the script itself, `sh` is set if it is not provided through Docker Compose.

sbaerlocher committed:
fix: feedback

sbaerlocher commented on commands/_internals/loadProjectDotdde.sh:
```diff
@@ -0,0 +1,57 @@
+        _logYellow "File $DDE_FILE not found. Continuing without loading .dde.yml."
+    else
+        # Read the version number from the .dde.yml file
+        local DDE_VERSION
```
Merged into a single line.
